bsp/src/mill/bsp/ModuleUtils.scala

```diff
@@ -112,15 +111,30 @@ object ModuleUtils {
     target
   }
 
-  def getMillBuildClasspath(evaluator: Evaluator, source: Boolean): Seq[String] = {
-    val all = Try(evaluator.rootModule.getClass.getClassLoader.asInstanceOf[SpecialClassLoader]).fold(
+  def getMillBuildClasspath(evaluator: Evaluator, sources: Boolean): Seq[String] = {
+    val classpath = Try(evaluator.rootModule.getClass.getClassLoader.asInstanceOf[SpecialClassLoader]).fold(
       _ => Seq.empty,
       _.allJars
     )
-    val filtered =
-      if (source) all.filter(url => isSourceJar(url))
+
+    val millVersion = BuildInfo.millVersion
+    val millJars = resolveDependencies(
+      Resolve.defaultRepositories,
+      depToDependency(_, BuildInfo.scalaVersion),
+      Agg(
+        ivy"com.lihaoyi::mill-main:$millVersion",
+        ivy"com.lihaoyi::mill-scalalib:$millVersion",
+        ivy"com.lihaoyi::mill-scalajslib:$millVersion",
+        ivy"com.lihaoyi::mill-scalanativelib:$millVersion"
+      ),
+      sources = sources
```

Comment:
I think these deps were never there before, at least all my experiments with mill-bsp resulted in missing mill API when editing `build.sc`. But, we should not bury such  a hardcoded list here. Instead, this is knowledge best located in the buildfile (`build.sc`). We could generate a Scala file there (`generateSources` or BuildInfo plugin) and use it here. We have this exactly hard-coded list also in `GenIdeaImpl`, and I think this is an opportunity to fix both places the right way.

Reply:
Excellent idea! Check out the new commit.